Create Event model

barangerbenjamin opened the pull request:
⚠️  [WIP] Do not merge ⚠️

barangerbenjamin commented:
Added `#published-btn` ID to the submit btn for publish an event
Added `#published-btn with` ```margin-left: 10px; ```to avoid two btn stuck together
Added `item-container__date` to display date of event in the lists of events
Removed useless commented lines on cities controller
Added events controller with full CRUD
Changed redirection after Auth to the first city available for a user instead of /admin plain
Added Event model
Added `published` column to Event table
Modified story model where company had the old column name for the company url
Modified the flash notice from every folder to a `shared` folder

@cedricmenteau It might be good to review the style changes as they might not be the best 😅

ssaunier commented on db/migrate/20180423153054_change_data_type_for_category.rb:
```diff
@@ -0,0 +1,5 @@
+class ChangeDataTypeForCategory < ActiveRecord::Migration[5.2]
+  def change
+    change_column :events, :category, "varchar[] USING (string_to_array(category, ','))"
```
@barangerbenjamin what is this type? An event has just one category.

barangerbenjamin replied:
@ssaunier Allow to be an array of string, like this 👇 
![capture d ecran 2018-04-24 a 12 00 29](https://user-images.githubusercontent.com/26072651/39183190-1dcf50ca-47b7-11e8-837f-0baf7601a600.png)

Otherwise, I have to hard code the possible categories of an event like this  👇 
```<%= f.input :category, collection: %w[Talk Workshop Demoday], label: "Category & template" %>```

With @cedricmenteau, we thought it would be cleaner to have this in the DB rather than being hard coded.

barangerbenjamin replied:
@ssaunier After discussing the issue with Sandrine, I reverted the migration but we thought : 
Wouldn't it be easier to have an `event_category` table with just a name column
Event ```has_one :event_category```
Category ```belongs_to :event```
That would allow us to do something like this 👇 
```<%= f.collection :category, label: "Category & template" %>```
🤔

ssaunier replied:
A bit overkill. Here we just have 3 types of event, it can be hard-coded as an array in `Event` and if we happen to have a new event type we'll update the code / redeploy.

barangerbenjamin replied:
Ok, I reverted to hard code 👌

ssaunier commented on app/views/admin/events/_form.html.erb:
```diff
@@ -0,0 +1,36 @@
+  <div class="admin-header">
+    <ul class="admin-header-breadcrumbs">
+      <li><%= link_to "Events", admin_city_events_path(params[:city_id]) %></li>
+      <% if params["action"] == "new" %>
```
It's always better to check at the record being used in the form:

```ruby
f.object.new_record?
```

barangerbenjamin replied:
Done, thanks 🙏

barangerbenjamin commented:
Modified Event controller to add the logic allowing dynamic links on index page (All, Talk, Demoday)
Modified set_city to allow it to work either with `find_by` OR `find` (depending of the action we get a `slug` or an ìd` in the params
Added ```f.object.new_record?``` for Event edit
added ```f.object.new_record?``` for Company Edit
Hide link to `My events` in the sidebar when `@city.nil`
Remove migration to `array: true` for `category`

ssaunier commented on app/controllers/admin/events_controller.rb:
```diff
@@ -0,0 +1,69 @@
+    def index
+      if params[:query].present?
+        @events = Event.where(city_id: @city).search_by_name(params[:query])
+      elsif params[:category] && params[:category] != ""
```
```ruby
params[:category].present?
```

barangerbenjamin replied:
✅

ssaunier commented on app/controllers/admin/events_controller.rb:
```diff
@@ -0,0 +1,69 @@
+    end
+
+    def update
+      params["commit"] == "Publish event" ? @event.published = true : @event.published = false
```
Relying on the button name is not a good practise as it would break the moment this button label is updated. Apparently you can pass a `value` in a button, see [this](https://stackoverflow.com/questions/4171664/html-submit-button-different-value-button-text)

barangerbenjamin replied:
✅

ssaunier replied:
For the value maybe use a `slug` like `publish` instead of `Publish event` !

ssaunier commented on app/controllers/users/omniauth_callbacks_controller.rb:
```diff
@@ -4,7 +4,7 @@ def kitt
     if @user
       sign_in @user, :event => :authentication
-      redirect_to admin_root_path
+      redirect_to admin_city_path(@user.cities.first.slug)
```
You need a `if / else` statement testing for `@user.cities.empty?`

barangerbenjamin replied:
✅ 
If user is admin and doesn't have any cities => redirected to admin_root_path
if user is admin and has cities => redirected to his first city ( but still has access to all cities)
if user is not admin but has cities => redirected to his first city (but no access to other cities)
if user is not admin and has no cities => redirect to www with flash notice explaining why

ssaunier commented on app/views/admin/events/_form.html.erb:
```diff
@@ -0,0 +1,37 @@
+    </ul>
+    <div class="admin-header-actions">
+      <% if f.object.new_record? %>
+        <%= f.button :submit, 'Create new event', class: 'admin-button admin-button-primary' %>
```
When creating an event, is it published? If not, the button should say "Save draft"

barangerbenjamin replied:
When creating a event, the default value for `published` is `false`
So I changed the text ✅

ssaunier commented on app/views/admin/events/_form.html.erb:
```diff
@@ -0,0 +1,37 @@
+      <% if f.object.new_record? %>
+        <%= f.button :submit, 'Create new event', class: 'admin-button admin-button-primary' %>
+      <% else %>
+        <%= f.button :submit, 'Save as draft', class: 'admin-button admin-button-primary' %>
```
Then this button should say "Update draft"

barangerbenjamin replied:
✅

ssaunier commented on app/views/admin/events/index.html.erb:
```diff
@@ -0,0 +1,72 @@
+      %>
+    <% end %>
+    <div class="admin-header-actions">
+      <%= link_to "Create a new event", new_admin_city_event_path, class: 'admin-button admin-button-primary' %>
```
Let's even be more "business-oriented":

```
Plan an event
```

ssaunier commented on db/schema.rb:
```diff
@@ -110,6 +110,23 @@
+    t.string "name"
+    t.text "description"
+    t.string "thumbnail"
+    t.datetime "occurs_on"
```
Should be named `starts_at` if it's a `datetime`. `_on` suffix is for `date` type.

barangerbenjamin replied:
✅ 
Also changed the form for it
Also changed the index for it

ssaunier commented on db/schema.rb:
```diff
@@ -110,6 +110,23 @@
+    t.datetime "created_at", null: false
+    t.datetime "updated_at", null: false
+    t.boolean "published", default: false
+    t.datetime "occurs_end"
```
That would be `ends_at`

barangerbenjamin replied:
✅ 
Also changed the form for it

ssaunier commented on spec/models/event_spec.rb:
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
```
Please remove boilerplate files.

barangerbenjamin replied:
✅ File deleted

ssaunier commented:
@barangerbenjamin Please solve conflicts 🙏

ssaunier commented:
(OK for pushing meetup / eventbrite later as we still have to sync with the marketing squad on how we'll use those tools with drivers)

ssaunier commented:
LGTM

barangerbenjamin commented:
@ssaunier Conflict solved, can you merge 🙏 ? Thanks
